core/src/main/java/org/apache/brooklyn/enricher/stock/Joiner.java

```diff
@@ -92,36 +96,48 @@ protected Object compute(SensorEvent<T> event) {
         Object result = null;
         if (v!=null) {
             if (v instanceof Map) {
-                v = ((Map<?,?>)v).values();
+                if (config().get(JOIN_MAP_ENTRIES)) {
+                    v = ((Map<?,?>) v).entrySet();
+                } else {
+                    v = ((Map<?,?>) v).values();
+                }
             }
             if (!(v instanceof Iterable)) {
                 LOG.warn("Enricher "+this+" received a non-iterable value "+v.getClass()+" "+v+"; refusing to join");
             } else {
                 MutableList<Object> c1 = MutableList.of();
-                Integer maximum = getConfig(MAXIMUM);
+                Integer maximum = config().get(MAXIMUM);
                 for (Object ci: (Iterable<?>)v) {
                     if (maximum!=null && maximum>=0) {
                         if (c1.size()>=maximum) break;
                     }
-                    c1.appendIfNotNull(Strings.toString(ci));
+                    if (ci instanceof Map.Entry) {
+                        String key = Strings.toString(((Map.Entry) ci).getKey());
+                        Object value = ((Map.Entry) ci).getValue();
+                        String keyValueSeparator = config().get(KEY_VALUE_SEPARATOR);
+                        if (value != null) {
+                            c1.append(String.format("%s%s%s", key, keyValueSeparator, Strings.toString(value)));
```

Comment:
add a `// TODO we might want to handle QUOTE=true specially for this case, to quote keys and values if needed` ?